Enhancement/efi shell interface

RenTrieu opened the pull request:
This PR will implement the wrappers for EFI Shell Protocol functions to address #448 . This PR does branch off of the changes made in #596 so big thanks to @timrobertsdev for laying the groundwork!

Currently, I've implemented wrappers and wrote tests for ``GetEnv()``, ``SetEnv()``, ``GetCurDir()`` and ``SetCurDir()``. My plan is to first write tests for and finish implementing ``Execute()``, ``CloseFile()``,  ``CreateFile()``, ``FindFiles()``, ``FindFilesInDir()``, and ``GetFileSize()`` since they have already been started. Then I can implement the rest in either this PR or subsequent follow up PRs (whichever is preferable).

I do have some questions about my ``get_env()`` implementation. UEFI's ``GetEnv()`` returns the value of a variable if its name is specified. However if its name is not specified, it returns all known variable names in a ``*const Char16`` where the names are delimited by a ``NULL`` and the end of the list is terminated by a double ``NULL``. 

My initial approach was to parse the ``*const Char16`` into a ``Vec`` in this case. I wrapped the return value in an Enum that can be unpacked into a single ``&CStr16`` or into a vector of ``&CStr16``s depending on which is expected. Is this approach okay? I was concerned that if I simply returned a ``&CStr16`` in the "name list" case that the true size of the string wouldn't be visible since the names are separated by ``NULL``. Also, is it okay for me to use ``alloc::vec::Vec``? I saw that a lot of other protocols don't use any ``Vec`` at all so I'm not clear on if there are some restrictions for doing so.

## Checklist
- [ ] Execute
- [ ] GetEnv
- [ ] SetEnv
- [ ] GetAlias
- [ ] SetAlias
- [ ] GetHelpText
- [ ] GetDevicePathFromMap
- [ ] GetMapFromDevicePath
- [ ] GetDevicePathFromFilePath
- [ ] GetFilePathFromDevicePath
- [ ] SetMap
- [ ] GetCurDir
- [ ] SetCurDir
- [ ] OpenFileList
- [ ] FreeFileList
- [ ] RemoveDupInFileList
- [ ] BatchIsActive
- [ ] IsRootShell
- [ ] EnablePageBreak
- [ ] DisablePageBreak
- [ ] GetPageBreak
- [ ] GetDeviceName
- [ ] GetFileInfo
- [ ] SetFileInfo
- [ ] OpenFileByName
- [ ] CloseFile
- [ ] CreateFile
- [ ] ReadFile
- [ ] WriteFile
- [ ] DeleteFile
- [ ] DeleteFileByName
- [ ] GetFilePosition
- [ ] SetFilePosition
- [ ] FlushFile
- [ ] FindFiles
- [ ] FindFilesInDir
- [ ] GetFileSize
- [ ] OpenRoot
- [ ] OpenRootByHandle
- [ ] ExecutionBreak
- [ ] Version Variables
- [ ] RegisterGuidName
- [ ] GetGuidName
- [ ] GetGuidFromName
- [ ] GetEnvEx

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -0,0 +1,584 @@
+    /// * `Some(Vec<env_names>)` - Vector of environment variable names
+    /// * `None` - Environment variable doesn't exist
+    #[must_use]
+    pub fn get_env<'a>(&'a self, name: Option<&CStr16>) -> Option<EnvOutput<'a>> {
```
I'd much rather have the high-level API to not replicate the low-level interface here. Instead, we can use the low-level interface to create an API similar to:
- [`std::env::var`](https://doc.rust-lang.org/std/env/fn.var.html)
- [`std::env::vars`](https://doc.rust-lang.org/std/env/fn.vars.html)

Then we also don't need the `EnvOutput` type effectively acting as a `Either` type.

phip1611 reviewed (changes requested):
Thanks for working on this! As a general rule of thumb: High-level wrappers should be close to the API of `std` and not replicate "ugly" or unusual details of the low-level API.

nicholasbishop commented:
Some general guidance:
1. Start with a PR that just updates `uefi-raw`. (As Philipp noted [here](https://github.com/rust-osdev/uefi-rs/pull/1679/files#r2115236234), the definition of the shell protocol that matches the C spec should go in uefi-raw. See for example [`RngProtocol`](https://github.com/rust-osdev/uefi-rs/blob/main/uefi-raw/src/protocol/rng.rs#L40). [api_guidelines](https://github.com/rust-osdev/uefi-rs/blob/main/uefi-raw/api_guidelines.md) has some notes about how to implement stuff in uefi-raw.)
2. Once that's merged, do small PRs that add ergonomic wrappers to the `uefi` crate. Small PRs will help get review done faster.
3. Regarding `Vec`, it's generally OK to use, but does need to be gated by the `alloc` feature. (I haven't looked at the code yet though, I'm just commenting generally.)

RenTrieu commented:
Thanks for the feedback! I will start with making the PR to update ``uefi-raw`` as Nicholas suggested, then come back to address the other comments.

nicholasbishop commented:
That sounds good, start with doing that in this PR. Once the changes are in this PR it's possible I might ask for it to be split up into more than one PR for review, but we can start with the assumption that one PR is OK.

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,141 @@
+    /// * `Some(Vec<env_names>)` - Vector of environment variable names
+    /// * `None` - Environment variable doesn't exist
+    #[must_use]
+    pub fn get_env<'a>(&'a self, name: Option<&CStr16>) -> Option<Vec<&'a CStr16>> {
```
Please don't export this unintuitive/unrusty UEFI API in this high-level wrapper. Instead, please aim to stay close to `std`. This means:
- https://doc.rust-lang.org/proc_macro/tracked_env/fn.var.html (with `&CStr16)
- https://doc.rust-lang.org/std/env/fn.vars.html )

RenTrieu replied:
Hi Philipp, thanks again for the resources. However, it is not clear to me what you are asking. Are you asking for ``get_env()`` to have a similar return type to the ``var()`` and ``vars()`` functions? Or are you asking for ``get_env()`` to be split up into two separate functions- one dedicated to returning the value of an environment variable and the other dedicated to returning the current environment variables?

phip1611 replied:
Sorry for the confusion!

> Or are you asking for get_env() to be split up into two separate functions- one dedicated to returning the value of an environment variable and the other dedicated to returning the current environment variables?

This.

RenTrieu replied:
No worries! I can see how having separate dedicated functions is more intuitive. In the spirit of mirroring ``std``, I have updated the protocol with these two new functions using the same naming convention as ``var`` and ``vars`` where:
``get_env()`` - retrieves the value of an environment variable
``get_envs()`` - retrieves the names of all environment variables

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,144 @@
+    ///
+    /// * `name` - The environment variable for which to set the value
+    /// * `value` - The new value of the environment variable
+    /// * `volatile` - Indicates whether or not the variable is volatile or
```
```suggestion
    /// * `volatile` - Indicates whether the variable is volatile or
```

phip1611 commented on uefi-test-runner/src/proto/shell.rs:
```diff
@@ -1,13 +1,182 @@
+    assert_eq!(cur_env_vec.count(), default_len);
+}
+
+/// Test ``get_cur_dir()`` and ``set_cur_dir()``
```
```suggestion
/// Test `get_cur_dir()` and `set_cur_dir()`
```

RenTrieu commented:
I've updated the PR to use the ``cstr16!`` macro instead of the buffers and also added a unit test for the ``Vars`` struct. Thank you for the tip on mocking the inner value with the ``Vec``, it was very helpful. Please let me know if the unit test looks alright.

phip1611 commented on uefi-test-runner/src/proto/shell.rs:
```diff
@@ -1,13 +1,168 @@
+    /* Test retrieving list of environment variable names */
+    let mut cur_env_vec = shell.get_envs();
+    assert_eq!(cur_env_vec.next().unwrap(), cstr16!("path"),);
+    assert_eq!(cur_env_vec.next().unwrap(), cstr16!("nonesting"),);
```
Where does `nonesting` come from?! Looks a little magic to me

phip1611 replied:
Interesting, it is pre-defined: https://uefi.org/sites/default/files/resources/UEFI_Shell_2_2.pdf

RenTrieu commented:
Commits have been squashed :)

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    ///   environment variable
+    /// * `None` - If environment variable does not exist
+    #[must_use]
+    pub fn get_env(&self, name: &CStr16) -> Option<&CStr16> {
```
In https://github.com/rust-osdev/uefi-rs/pull/1679#discussion_r2128040189, it was suggested to use the same name as `std`: `var`. I think we should do that here, for several reasons:
* Consistency in APIs is good; anyone who is familiar with `std::env` will instinctively know how to use this API
* Per https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter, Rust APIs generally should not use a `get_` prefix
* It's clearer to say you are getting the variable (noun) rather than the environment (which can be a noun, but in "environment variable" it's an adjective).

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    ///
+    /// # Returns
+    ///
+    /// * `Vec<env_names>` - Vector of environment variable names
```
Out of date, it returns `Vars` now. You can probably drop the `Returns` section here though, it's redundent with the top-level description.

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+        }
+    }
+
+    /// Gets the list of environment variables
```
"Gets an iterator over the names of all environment variables."

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    ///
+    /// * `Vec<env_names>` - Vector of environment variable names
+    #[must_use]
+    pub fn get_envs(&self) -> Vars {
```
I think `var_names` would be a better name for this function (see comment on `get_env` for details). However, even better might be to make it match `std` and have `Vars` return an iterator over both names _and values_ (and then the method can be named simply `vars`). Is that doable?

phip1611 replied:
Yes, sounds reasonable. High level abstractions should be close to known interfaces (such as std) and not replicate low-level details

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    ///
+    /// * `Status::SUCCESS` - The variable was successfully set
+    pub fn set_env(&self, name: &CStr16, value: &CStr16, volatile: bool) -> Status {
+        let name_ptr: *const Char16 = core::ptr::from_ref::<CStr16>(name).cast();
```
Instead of `core::ptr::from_ref`, you can use `name.as_ptr()` (and ditto for value below).

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    /// * `Some(cwd)` - CStr16 containing the current working directory
+    /// * `None` - Could not retrieve current directory
+    #[must_use]
+    pub fn get_cur_dir(&self, file_system_mapping: Option<&CStr16>) -> Option<&CStr16> {
```
Rename to `current_dir` to match [std](https://doc.rust-lang.org/std/env/fn.current_dir.html).

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    /// * `None` - Could not retrieve current directory
+    #[must_use]
+    pub fn get_cur_dir(&self, file_system_mapping: Option<&CStr16>) -> Option<&CStr16> {
+        let mapping_ptr: *const Char16 = file_system_mapping.map_or(ptr::null(), |x| (x.as_ptr()));
```
nit: parens not needed around `x.as_ptr()`. Might be clearer to write as `map_or(ptr::null(), CStr16::as_ptr)`.

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    /// # Errors
+    ///
+    /// * `Status::EFI_NOT_FOUND` - The directory does not exist
+    pub fn set_cur_dir(&self, file_system: Option<&CStr16>, directory: Option<&CStr16>) -> Status {
```
`set_current_dir` to match [std](https://doc.rust-lang.org/std/env/fn.set_current_dir.html)

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+    /// # Returns
+    ///
+    /// * `Status::SUCCESS` - The variable was successfully set
+    pub fn set_env(&self, name: &CStr16, value: &CStr16, volatile: bool) -> Status {
```
Here and on other methods, return `Result` (https://docs.rs/uefi/latest/uefi/type.Result.html) instead of a raw `Status`. This is more convenient for users since it works with `?`.

nicholasbishop replied:
`set_var` to match [std](https://doc.rust-lang.org/std/env/fn.set_var.html)

nicholasbishop commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -2,12 +2,208 @@
+            Some(CStr16::from_ptr(cur_start))
+        }
+    }
+}
```
I think the implementation could be simplified, something like this:
```rust
    fn next(&mut self) -> Option<Self::Item> {
        let s = unsafe { CStr16::from_ptr(self.inner) };
        if s.is_empty() {
            None
        } else {
            self.inner = unsafe { self.inner.add(s.num_chars() + 1) };
            Some(s)
        }
    }
```

phip1611 commented:
> Thank you for the feedback. I'm starting to understand the goal of abstracting the lower level implementation in ways that are useful for the user 

Perfect! Thanks for your patience and valuable contributions :)

> For returning Result as opposed to Status, is it alright to simply call the to_result() method and return?

Yes, perfectly fine, and also done in other parts of the code base

> As for changing Vars to return both the name and value of the environment variables, is it fine to use something from the alloc crate like Vec or Box for this

Yes, that's fine. So far, we didn't have a clear guidline. IMHO, `uefi` in the early days tried to avoid `alloc` whenever possible, but over time, `alloc` was mandatory for more and more parts of the crate.

Now, IMHO, the guideline is: prevent `alloc` when it is easily doable. Otherwise, go with the convenience of `alloc`

RenTrieu commented:
> Now, IMHO, the guideline is: prevent ``alloc`` when it is easily doable. Otherwise, go with the convenience of ``alloc``

In this case, would be it be okay to simply store the vector of values as one of the fields of ``Var``? I was thinking of something like:
```rust
pub struct Vars<'a> {
    /// Char16 containing names of environment variables
    names: *const Char16,
    /// Vec containing values of environment variables
    values: Vec::<Option<&'a CStr16>>,
    /// Current position in vec
    cur_pos: usize,
    /// Placeholder to attach a lifetime to `Vars`
    placeholder: PhantomData<&'a CStr16>,
}
```

nicholasbishop commented:
Hi, just getting back to reviewing this. Could you create a new PR with the current_dir/set_current_dir changes, so that this one just has var/vars/set_var? Getting the per-PR changes smaller will help me review it more quickly.

phip1611 commented:
Hey @RenTrieu  I know this is taking quite long, but we are doing very good and approaching the finishing line! Could you please rebase your branch? Then we can talk about a pragmatic way forward to merge more functionality

RenTrieu commented:
@phip1611 I've rebased my branch. Please let me know if there are still any improvements you think we can make. By the way, as long as it is okay with you and Nicholas, I don't mind the pace that the PR process has been going at. Should we discuss the approach to future functionality in this thread or would #448 be more appropriate?

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -54,4 +55,82 @@ impl Shell {
+    /// * `Some(Vec<env_names>)` - Vector of environment variable names
+    /// * `None` - Environment variable doesn't exist
+    #[must_use]
+    pub fn get_env<'a>(&'a self, name: Option<&CStr16>) -> Option<Vec<&'a CStr16>> {
```
I thought you changed that? [Context.](https://github.com/rust-osdev/uefi-rs/pull/1679/commits/6547011b76e5c58a25233dc00f75f133a7cc398c#r2129055664)

At this point, I expected to have
- `var(name: &CStr16>) -> Option<&CStr16> {}`
- `vars() -> Vars` (Iterator over all vars)
- `set_var(name: &CStr16>, value: &CStr16) -> Option<&CStr16> {}`

All three functions should work akin to [this](https://doc.rust-lang.org/stable/std/env/fn.var.html), [this](https://doc.rust-lang.org/stable/std/env/fn.vars.html), and [this](https://doc.rust-lang.org/stable/std/env/fn.set_var.html).

phip1611 commented on uefi-test-runner/src/proto/shell.rs:
```diff
@@ -100,6 +100,65 @@ pub fn test_current_dir(shell: &ScopedProtocol<Shell>) {
+    assert!(!found_var);
+    let cur_env_vec = shell.vars();
+    let mut found_var = false;
+    for (env_var, _) in cur_env_vec {
```
nit: you could use `cur_env_vec.iter().any()` here I think

phip1611 commented on uefi-test-runner/src/proto/shell.rs:
```diff
@@ -100,6 +100,65 @@ pub fn test_current_dir(shell: &ScopedProtocol<Shell>) {
+
+    let cur_env_vec = shell.vars();
+    let mut found_var = false;
+    for (env_var, _) in cur_env_vec {
```
nit: you could use `cur_env_vec.iter().any()` here I think

RenTrieu commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -13,6 +15,45 @@ use uefi_raw::protocol::shell::ShellProtocol;
 pub struct Shell(ShellProtocol);
 
+/// Trait for implementing the var function
+pub trait ShellVar {
```
Hi Philipp, yes it is for testability. I needed some way to pass the ``ShellMock`` object in as a protocol when creating a ``Vars`` iterator in the unit tests and the trait was the best way I could come up with. Thank you for the feedback, I will rename it appropriately.

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -54,4 +95,159 @@ impl Shell {
+    }
+
+    impl<'a> ShellMock<'a> {
+        fn new(names: Vec<&'a CStr16>, values: Vec<&'a CStr16>) -> ShellMock<'a> {
```
At least

```suggestion
        fn new(names: &[&'a CStr16], values: &[&'a CStr16]) -> ShellMock<'a> {
```

you should also assert that both's len() equals.

But **much better** would be:


```suggestion
        fn new(pairs: impl IntoIterator<Item = (&CStr16, &CStr16)>) -> ShellMock<'a> {
```

as this is your unit test where you have control over. Your initial strategy is way too complicated; it can be done much simpler :)

phip1611 commented on uefi/src/proto/shell/mod.rs:
```diff
@@ -54,4 +95,159 @@ impl Shell {
+
+        // Multiple environment variables in Vars
+        let mut vars_mock = Vec::<u16>::new();
+        vars_mock.push(b'f' as u16);
```
Ahh yeah, my bad. Then the original approach is fine

phip1611 commented:
Thanks for your patience! I've added two commits and the thing will be merged in a couple of seconds.

I hope you learned something and future upstreaming efforts will be much easier for you!
